compiler/rustc_mir/src/transform/lower_slice_len.rs

```diff
@@ -0,0 +1,144 @@
+
+pub struct LowerSliceLenCalls;
+
+impl<'tcx> MirPass<'tcx> for LowerSliceLenCalls {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        lower_slice_len_calls(tcx, body)
+    }
+}
+
+pub fn lower_slice_len_calls<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    // FIXME: (shamatar) Can be tuned later
+    // if tcx.sess.mir_opt_level() < 3 {
+    //     return;
+    // }
+
+    // Optimization is once per basic block, so we just visit terminators
+    let languate_items = tcx.lang_items();
+    let slice_lang_item_def_id = if let Some(slice_lang_item) = languate_items.slice_impl() {
+        slice_lang_item
+    } else {
+        // there is no language item to compare too :)
+        return;
+    };
+    let len_symbol = rustc_span::Symbol::intern("len");
+
+    let mut visitor = LowerLenCallVisitor { tcx, len_symbol, slice_lang_item_def_id };
+    visitor.visit_body(body);
+}
+
+struct LowerLenCallVisitor<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    len_symbol: rustc_span::Symbol,
+    slice_lang_item_def_id: DefId,
+}
+
+struct PatchInformation<'tcx> {
+    add_statement: Statement<'tcx>,
+    new_terminator_kind: TerminatorKind<'tcx>,
+}
+
+impl<'tcx> MutVisitor<'tcx> for LowerLenCallVisitor<'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_basic_block_data(&mut self, _block: BasicBlock, data: &mut BasicBlockData<'tcx>) {
+        // We do not need to call super here, so just work with it
+
+        let mut patch_found: Option<PatchInformation<'_>> = None;
+
+        let terminator = data.terminator();
+        match &terminator.kind {
+            TerminatorKind::Call {
+                func: Operand::Constant(box func_sig),
+                args,
+                destination: Some((dest, bb)),
+                cleanup: None,
+                from_hir_call: true,
+                ..
+            } => {
+                // some heuristics for fast rejection
+                if args.len() != 1 {
+                    return;
+                }
+                let (arg, _is_move) = match &args[0] {
+                    Operand::Copy(arg) => (*arg, false),
+                    Operand::Move(arg) => (*arg, true),
+                    _ => return,
+                };
+                let sig =
+                    if let Some(sig) = func_sig.literal.const_for_ty() { sig } else { return };
+                // match over function signature
+                match sig {
+                    ty::Const {
+                        ty,
+                        val:
+                            ty::ConstKind::Value(rustc_middle::mir::interpret::ConstValue::Scalar(
+                                rustc_middle::mir::interpret::Scalar::Int(s),
+                            )),
+                    } => {
+                        // more early rejections
+                        if s != &ty::ScalarInt::ZST || !ty.is_fn() {
+                            return;
+                        }
```

Comment:
I don't think this is actually faster than immediatly checking `sig.ty` without matching on `sig` and checking for ZSTs.